mythril_core/src/acpi.rs

```diff
@@ -0,0 +1,188 @@
+#![deny(missing_docs)]
+use crate::error::{Error, Result};
+use byteorder::{ByteOrder, NativeEndian};
+use core::fmt;
+use core::ops::Range;
+
+/// Extented BIOS Data Area Start Address
+const EXTENDED_BIOS_DATA_START: usize = 0x000040e;
+/// Extented BIOS Data Area End Address
+const EXTENDED_BIOS_DATA_END: usize = EXTENDED_BIOS_DATA_START + 1024;
+/// Main BIOS Data Area Start Address
+const MAIN_BIOS_DATA_START: usize = 0x000e0000;
+/// Main BIOS Data Area End Address
+const MAIN_BIOS_DATA_END: usize = 0x000fffff;
+/// Well Known RSDP Signature
+const RSDP_SIGNATURE: &[u8; 8] = b"RSD PTR ";
+
+/// Offsets from [ACPI § 5.2.5.3]
+///
+/// [ACPI § 5.2.5.3]: https://uefi.org/sites/default/files/resources/ACPI_6_3_May16.pdf
+mod offsets {
+    use super::*;
+    /// Well known bytes, "RST PTR ".
+    pub const SIGNATURE: Range<usize> = 0..8;
+    /// Checksum of the fields defined by ACPI 1.0.
+    pub const CHECKSUM: usize = 8;
+    /// OEM-supplied ID string.
+    pub const OEMID: Range<usize> = 9..15;
+    /// Revision of the structure. Zero for ACPI 1.0 and two for ACPI 2.0.
+    pub const REVISION: usize = 15;
+    /// 32-bit physical address of the RSDT.
+    pub const RSDT_ADDR: Range<usize> = 16..20;
+    /// Length of the table in bytes (ACPI 2.0 only).
+    pub const LENGTH: Range<usize> = 20..24;
+    /// 64-bit physical address of the XSDT (ACPI 2.0 only).
+    pub const XSDT_ADDR: Range<usize> = 24..32;
+    /// Checksum of entire structure (ACPI 2.0 only).
+    pub const EXT_CHECKSUM: usize = 32;
+    /// Size of the RSDP, including the reservied field.
+    pub const SIZE: usize = 36;
+}
+
+/// The Root System Descriptor Pointer (RSDP)
+pub struct RSDP {
```

Comment:
Isn't this a sort of natural place for an enum? One for each revision?

Reply:
We've done things kind of like this in the past https://github.com/jcreekmore/dmidecode/blob/master/src/lib.rs. Though I'm not sure that's exactly the interface you'd want to mimic.

Reply:
Yeah, good call